rmm/monitor/src/rmm/granule/mod.rs

```diff
@@ -59,33 +61,124 @@ impl GranuleState {
+///   - this can be seen as a typical transaction that defines a way to access granule. (granule transaction)
+///   - to open a transaction, you can use either `get_granule` or `get_granule_if` or `set_state_and_get_granule`.
+///   - to modify a granule and close a transaction, `set_granule` can be used.
+///
+/// - Safety:
+///   - In the above usage, `set_granule()` takes &entry::Inner as input, which can only be acquired by `get_granule_if!`.
+///     This means that you can access granule only after `get_granule_if!`, and this is ensured by this API design.
+///   - TODO(Optional): currently, handling a granule transaction is in charge of developers.
+///           e.g., (1) they should invoke `set_granule()` at the right moment, (2) `EntryGuard` still lives after `set_granule()`.
+///           we might be able to free them from it via Rust features.
+///   - TODO(Must): currently, these APIs do not involve in mapping some address into RMM page table.
+///           callers must map some physical address, if needed, before/after using these APIs.
+///           we need to define and use a secure interface to map some address and do set_granule().
+///
+/// - Note:
+///   - why is it using macros, not functions?
+///     writing a function that returns `EntryGuard<>` requires cloning EntryGuard. To get around this, macros are currently used.
+///     we might be able to figure out a better way to define `get_granule_*!` macros.
+
+/// get_granule!(addr: a physical address)
+/// - when success, returns `EntryGuard<entry::Inner>` allowing an access to "Granule".
+#[macro_export]
+macro_rules! get_granule {
+    ($addr:expr) => {{
+        {
+            use crate::mm::address::PhysAddr;
+            use crate::mm::error::Error as MmError;
+            use crate::mm::page::Page;
+            use crate::mm::page_table::{self, PageTableMethods};
+            use crate::rmm::granule::translation::{GranuleSize, GRANULE_STATUS_TABLE};
+            use crate::rmm::granule::{to_rmi_result, validate_addr};
+
+            let guard_res = if let Some(gst) = unsafe { &mut GRANULE_STATUS_TABLE } {
+                if !validate_addr($addr) {
+                    return to_rmi_result(Err(MmError::MmInvalidAddr));
+                }
+                let pa = Page::<GranuleSize, PhysAddr>::including_address(PhysAddr::from($addr));
+                let guard_res = gst.root_pgtlb.entry(pa, |e| page_table::Entry::lock(e));
+                let guard = match guard_res {
+                    Ok(g) => g,
+                    Err(e) => return to_rmi_result(Err(e)),
+                };
+                guard
+            } else {
+                return to_rmi_result(Err(MmError::MmErrorOthers));
+            };
+
+            let guard = match guard_res {
+                Some(g) => g,
+                _ => return to_rmi_result(Err(MmError::MmErrorOthers)),
+            };
+            guard
```

Comment:
Could we return Result<Guard, MMError> on `get_granule` and `get_granule_if`?
If it is changed, we can use like `let granule = get_granule!()?;`

```suggestion
            if !validate_addr($addr) {
                    Err(MmError::MmInvalidAddr);
            } else {
                 if let Some(gst) = unsafe { &mut GRANULE_STATUS_TABLE } {
                     let pa = Page::<GranuleSize, PhysAddr>::including_address(PhysAddr::from($addr));
                     match gst.root_pgtlb.entry(pa, |e| page_table::Entry::lock(e)).map(|g|  .. Ok(g) )
            } else {
                Err(MmError::MmErrorOthers))
            }
```

Reply:
`let granule = get_granule!()?;`  --> I really love it... many thanks. I'll do it.

Reply:
Reflected: